csharp/src/Microsoft.ML.OnnxRuntime/DisposableNamedOnnxValue.cs

```diff
@@ -78,8 +78,7 @@ private DisposableNamedOnnxValue(string name, Object value, NativeMemoryHandler
         /// <param name="onnxValue"></param>
         /// <param name="pinnedMemoryHandle"></param>
         /// <param name="disposeOnnxValueAfterUse"></param>
-        internal override void ToNativeOnnxValue(out IntPtr onnxValue, 
-                                                 out MemoryHandle pinnedMemoryHandle)
+        internal override void ToNativeOnnxValue(out IntPtr onnxValue, out MemoryHandle pinnedMemoryHandle, out bool disposeOnnxValueAfterUse)
```

Comment:
the param `disposeOnnxValueAfterUse` is added back because it resolved both @hariharans29 and @jignparm 's concerns
    - a boolean value to indicate Run()'s behavior is concept wise correct. the code relies on class's behavior rather than class's name to figure out the correct behavior
    - the worry of making interface more complicated is not necessary. first this is a internal scope, which does not affect public interface; second the core logic is moved to a helper class and there is no this flag.

Now it's clean: NamedOnnxValue (and its subclass) just popup the correct handle and a boolean value to indicate how these resources to be used and released; the helper class focus on how to create native resources .

Reply:
Thanks, I am okay with this. I went per @jignparm 's review, but not relying on class name does seem cleaner to me in general.

Reply:
The 'disposeOnnxValueAfterUser' parameter in the ToNativeOnnxValue() method makes it seem like this method has knowledge of whether or not to dispose the onnx values. The onnx values  should **always** be disposed, but the only  question is at what time? Before session.run(), after session.run(), when the user is done with an object, or at some other time?

The 2 classes 'NamedOnnxValue' and 'DisposableNamedOnnxValue' giving guidance to the caller function that the onnx values should be disposed  seems  bit weird  -- they should just return a handle to the onnx value. That would keep the  ToNativeOnnxValue API clean, and push off the disposal logic to calling functions.